cmd/pj-rehearse/main.go

```diff
@@ -142,6 +178,7 @@ func main() {
 		}
 	}
 
+	defer git(o.candidatePath, []string{"checkout", jobSpec.Refs.Pulls[0].Ref})
```

Comment:
why?

Reply:
It was annoying that the repository's `HEAD` ended up detached every time I executed it.

I haven't taken the time to figure out why yet.

Reply:
That happens because we use SHAs instead of refs when we checkout. From `git`'s man page:

```
       --detach
           Rather than checking out a branch to work on it, check out a commit
           for inspection and discardable experiments. This is the default
           behavior of "git checkout <commit>" when <commit> is not a branch
           name.
```

While this doesn't matter in the CI job, it is annoying when executed in a local repository, specially when iterating on a job — the HEAD is left pointing to the right commit, but detached. Was there any reason to choose one rather than the other? If there was, I think the current solution of `defer`ring a `git checkout` to restore the original HEAD is fine.

Reply:
> While this doesn't matter in the CI job, it is annoying when executed in a local repository, specially when iterating on a job — the HEAD is left pointing to the right commit, but detached. Was there any reason to choose one rather than the other? 

I think yes. My understanding is that when you run in a decorated Prowjob, you do not have `jobSpec.Refs.Pulls[0].Ref` checked out, you have it merged with the base.

Reply:
Right, looking at the code, it is all done on top of the base ref.

I think `defer`ing a checkout is the simplest solution here. I have moved it inside the `if o.local` conditional, since that is not necessary in CI.

Reply:
Creating named refs gets real messy real quick in CI -- you have to worry about collisions, etc. The CI knows exactly what it wants to be doing with SHAs and can do it more cleanly that way.